gps_ubxm8_i2c/gps.cpp

```diff
@@ -0,0 +1,117 @@
+ *
+ * There are also special return types for different types of error, that are hopefully verbose enough to explain themselves.
+ */
+const GPS::PollResult GPS::pollUpdate(I2C_HandleTypeDef* i2c) {
+    if (state == GPS::State::REQUEST_NOT_SENT) {
+        uint8_t message[4] = {0x01, 0x07, 0x00, 0x00};
+        sendUBX(message, 4, &hi2c3);
+        state = GPS::State::POLLING_RESPONSE;
+    }
+
+    if (state == GPS::State::POLLING_RESPONSE) {
+        uint8_t lenBytes[2];
+        uint16_t dataLen = 0;
+        uint8_t buffer[I2C_BUFFER_SIZE];
+
+        HAL_StatusTypeDef dataLenReadStatus = HAL_I2C_Mem_Read(i2c, 0x42 << 1, 0xFD, 1, lenBytes, 2, 100);
+        if (dataLenReadStatus != HAL_OK) {
+            return GPS::PollResult::DATA_LEN_POLL_FAILED;
+        }
+
+        dataLen = lenBytes[0] << 8 | lenBytes[1];
+        if (dataLen == 0) {
+            return GPS::PollResult::NO_DATA;
+        }
+
+        if (dataLen > I2C_BUFFER_SIZE) {
+            dataLen = I2C_BUFFER_SIZE;
+        }
+        HAL_StatusTypeDef rcvStatus = HAL_I2C_Master_Receive(i2c, 0x42 << 1, buffer, dataLen, 100);
+        if (rcvStatus != HAL_OK) {
+            return GPS::PollResult::DATA_RECEIVE_I2C_FAILED;
+        }
+
+        if (packetReader.isInProgress()) {
+            for (uint16_t i = 0; i < dataLen; i++) {
+                UBXPacketUpdateResult res = packetReader.update(buffer[i]);
+                if (res == UBXPacketUpdateResult::CHECKSUM_FAILED) {
+                    packetReader.reset();
+                    return GPS::PollResult::DATA_RECEIVE_CHECKSUM_FAILED;
+                }
+                if (packetReader.isComplete()) {
+                    state = GPS::State::RESPONSE_READY;
+                    return GPS::PollResult::POLL_JUST_FINISHED;
+                }
+            }
+            return GPS::PollResult::RECEIVE_IN_PROGRESS;
+        } else {
+            for (uint16_t i = 0; i < dataLen; i++) {
+                if (buffer[i] == 0xB5 && buffer[i + 1] == 0x62) {
+                    i += 2;  // skip the header
+                    while (i < dataLen && !packetReader.isComplete()) {
+                        UBXPacketUpdateResult res = packetReader.update(buffer[i]);
+                        if (res != UBXPacketUpdateResult::UPDATE_OK) {
+                            packetReader.reset();
+                            return GPS::PollResult::DATA_RECEIVE_CHECKSUM_FAILED;
+                        }
+                        i++;
+                    }
+                }
+            }
+
+            if (!packetReader.isInProgress()) {
+                return GPS::PollResult::NO_UBX_DATA;
+            }
+
+            if (packetReader.isComplete()) {
+                state = GPS::State::RESPONSE_READY;
+                return GPS::PollResult::POLL_JUST_FINISHED;
+            }
+            return GPS::PollResult::RECEIVE_IN_PROGRESS;
+        }
+    }
+    return GPS::PollResult::POLL_ALREADY_FINISHED;
+}
+
+/**
```

Comment:
use javadoc/doxygen style comments w/ `@brief` `@param` and `@retval`
VSCode automatically parses them and makes the mouse hover popup a lot nicer

also by convention move the general method comments to header file